Mathlib/Data/Nat/Factorial/Basic.lean

```diff
@@ -457,4 +457,67 @@ lemma two_pow_mul_factorial_le_factorial_two_mul (n : ℕ) : 2 ^ n * n ! ≤ (2
       Nat.mul_le_mul_left _ (Nat.pow_le_pow_left (le_add_left _ _) _)
     _ ≤ _ := Nat.factorial_mul_pow_le_factorial
 
+
+/-!
+# Factorial via binary splitting.
+
+We prove this is equal to the standard factorial and mark it `@[csimp]`.
+
+We could proceed further, with either Legendre or Luschny methods.
+-/
+
+/-!
+This is the highest factorial I can `#eval` using the naive implementation without a stack overflow:
+```
+/-- info: 114716 -/
+#guard_msgs in
+#eval 9718 ! |>.log2
+```
+
+We could implement a tail-recursive version (or just use `Nat.fold`),
+but instead let's jump straight to binary splitting.
+-/
+
+/-- Factorial implemented using binary splitting. -/
+def factorialBinarySplitting (n : Nat) : Nat :=
+  prodRange 1 (n + 1)
+where
+  /-- `prodRange lo hi` is the product of the range `lo` to `hi`, computed by binary splitting.-/
+  prodRange (lo hi : Nat) : Nat :=
```

Comment:
Perhaps 
```suggestion
  prodRange (lo hi : Nat) (h : lo ≤ hi): Nat :=
```
to eliminate a branch

Reply:
Ah, I guess that doesn't help much unless you move the branch outside the recursion

Reply:
I've moved the `if` outwards.